File: USACO/castle/Castle.java
/*
ID: kapiley1
PROG: castle
LANG: JAVA
*/

import java.io.*;
class WALL
{
	public boolean W,N,E,S;
}
 class castle
{
	static int M,N,MAX,MAXR,NCOMP=1,graph[][]=new int[53][53],comp[][]=new int[53][53],count[]=new int[2600];
	static WALL getWalls(int n)
	{
		WALL w=new WALL();
		if(n>=8)
		{n-=8;w.S=true;}
		if(n>=4)
		{n-=4;w.E=true;}
		if(n>=2)
		{n-=2;w.N=true;}
		if(n>=1)
		{n-=1;w.W=true;}
		return w;
	}
	static void floodfill()
	{
		int nvis=0;
		do
		{
			nvis=0;
			
			for(int i=0;i<N;i++)
			for(int j=0;j<M;j++)
			{
				if(comp[i][j]==-2)
				{
					nvis++;
					
					comp[i][j]=NCOMP;
					count[NCOMP]++;
					WALL w=getWalls(graph[i][j]);
					
					if(!w.W && comp[i][j-1]==-1)
					{
						comp[i][j-1]=-2;
					}
					if(!w.N && comp[i-1][j]==-1)
					{
						comp[i-1][j]=-2;
					}
					if(!w.E && comp[i][j+1]==-1)
					{
						comp[i][j+1]=-2;
					}
					if(!w.S && comp[i+1][j]==-1)
					{
						comp[i+1][j]=-2;
					}
				}
			}
		
		}while(nvis>0);
	}
	static void components()
	{
		for(int i=0;i<N;i++)
		for(int j=0;j<M;j++)
		{
			if(comp[i][j]==-1)
			{
				comp[i][j]=-2;
				floodfill();
				NCOMP++;
			}
		}
	}
	public static void main(String args[]) throws IOException
	{
		BufferedReader bf=new BufferedReader(new FileReader("castle.in"));
		PrintWriter pw=new PrintWriter(new File("castle.out"));
		
		String line=bf.readLine();
		String vals[]=line.split(" ");
		M=Integer.parseInt(vals[0]);
		N=Integer.parseInt(vals[1]);
		//input
		for(int i=0;i<N;i++)
		{
			line=bf.readLine();
			vals=line.split(" ");
			for(int j=0;j<M;j++)
			{
			graph[i][j]=Integer.parseInt(vals[j]);
			comp[i][j]=-1;
			}
		}
		//calculate components
		components();
		//largest room
		MAXR=1;
		for(int i=0;i<NCOMP;i++)
		{
			if(count[i]>MAXR)MAXR=count[i];
		}
		//remove wall
		MAX=1;
		int X=1,Y=1;
		WALL d=new WALL();
		d.E=d.N=false;
		for(int j=0;j<M;j++)
		for(int i=N-1;i>=0;i--)
		{
			int n;
			WALL w=new WALL();
			w=getWalls(graph[i][j]);
			if(i>0 && comp[i][j]!=comp[i-1][j])
			{
				n=count[comp[i][j]]+count[comp[i-1][j]];
				if(n>MAX)
				{
					MAX=n;
					X=i;
					Y=j;
					d.N=true;
					d.E=false;
				}
			}
			if(j<M-1 && comp[i][j]!=comp[i][j+1])
			{
				n=count[comp[i][j]]+count[comp[i][j+1]];
				if(n>MAX)
				{
					MAX=n;
					X=i;
					Y=j;
					d.N=false;
					d.E=true;
				}
			}
		}
		
		
		pw.write(NCOMP-1+"\n");
		pw.write(MAXR+"\n");
		pw.write(MAX+"\n");
		pw.write((X+1)+" "+(Y+1)+" "+(d.E?"E":"N")+"\n");
		
		pw.close();
	}
}

File: USACO/castle/castle.cpp
/*
ID: kapiley1
PROG: castle
LANG: C++
*/
#include<stdio.h>
int graph[53][53],comp[53][53],count[2600]={0};
int M,N;
int compNum=1;
class walls
{
	public:
	bool W=0,N=0,E=0,S=0;
	walls(int n)
	{
		if(n>=8)
		{
			S=1;
			n-=8;
		}
		if(n>=4)
		{
			E=1;
			n-=4;
		}
		if(n>=2)
		{
			N=1;
			n-=2;
		}
		if(n>=1)
		{
			W=1;
			n-=1;
		}
	}
	void setE()
	{
		E=1;
		N=0;
	}
	void setN()
	{
		N=1;
		E=0;
	}
	bool getE()
	{
		return E;
	}
	bool getN()
	{
		return N;
	}
};
void floodfill(int X,int Y)
{
	int numV=0;
	do
	{
		numV=0;
		for(int i=0;i<N;i++)
		for(int j=0;j<M;j++)
		{
			if(comp[i][j]==-2)
			{
				
				comp[i][j]=compNum;
				count[compNum]++;
				walls e(graph[i][j]);
				if(!e.N && comp[i-1][j]==-1)comp[i-1][j]=-2;
				if(!e.W && comp[i][j-1]==-1)comp[i][j-1]=-2;
				if(!e.S && comp[i+1][j]==-1)comp[i+1][j]=-2;
				if(!e.E && comp[i][j+1]==-1)comp[i][j+1]=-2;
				numV++;
			}
		}
	}while(numV>0);
}
void components()
{
		for(int i=0;i<N;i++)
		for(int j=0;j<M;j++)
		{
			if(comp[i][j]==-1)
			{
				comp[i][j]=-2;
				floodfill(i,j);
				compNum++;
			}
		}
}
int main()
{
	freopen("castle.in","r",stdin);
	freopen("castle.out","w",stdout);
	
	scanf("%d%d",&M,&N);
	for(int i=0;i<N;i++)
	for(int j=0;j<M;j++)
	{
		scanf("%d",&graph[i][j]);
		comp[i][j]=-1;
	}
	components();
	
	/*for(int i=0;i<N;i++)
	for(int j=0;j<M;j++)
	printf("%d%c",comp[i][j],j<M-1?' ':'\n');
	*/
	printf("%d\n",compNum-1);
	
	int MAXR=0;
	for(int i=0;i<compNum;i++)
	{
		if(MAXR<count[i])
		MAXR=count[i];
	}
	
	int MAX=-1,X=-1,Y=-1;
	walls w(0);
	int c=0;
	for(int j=0;j<M;j++)
	for(int i=N-1;i>=0;i--)
	{
		if(i>0 && comp[i][j]!=comp[i-1][j])
		{
			c=count[comp[i][j]]+count[comp[i-1][j]];
			if(c>MAX)
			{
				MAX=c;
				X=i;
				Y=j;
				w.setN();
			}
		}
		
		if(j<M-1 && comp[i][j]!=comp[i][j+1])
		{
			c=count[comp[i][j]]+count[comp[i][j+1]];
			if(c>MAX)
			{
				MAX=c;
				X=i;
				Y=j;
				w.setE();
			}
		}
		
	}
	printf("%d\n%d\n%d %d %c\n",MAXR,MAX,X+1,Y+1,w.E?'E':'N');
	return 0;
}

